### PR TITLE
More parallel synchronization assertions

### PR DESCRIPTION
**Pull request by roystgnr**
I'm not sure whether the hung jobs @YaqiWang has been seeing are actually due to loss of sync in libMesh/MOOSE code, rather than something lower-level, but it is a bit embarrassing that we're seeing a loss of sync in devel mode runs that should have been caught with an assertion and is instead manifesting as a hang.

### CODE REVIEW

File: src/numerics/petsc_vector.C [hunk 1]
```diff
@@ -353,6 +375,9 @@ void PetscVector<T>::add (const T a_in, const NumericVector<T> & v_in)
 
   PetscErrorCode ierr = VecAXPY(_vec, a, v->vec());
   LIBMESH_CHKERR(ierr);
+
+  libmesh_assert(this->comm().verify(this->type()));
```
Comment (lindsayad) [inline src/numerics/petsc_vector.C:379]: nice

Comment (roystgnr) [reply src/numerics/petsc_vector.C:379]: Oh, but hell, I meant to put that in `add_vector()` too.  After looking at the MOOSE patch again I don't think this is the problem, but if it is let's find it at its root.

**Comment by roystgnr** [comment]
This isn't going to work without a custom OpFunction, huh?  Okay, that's doable.

**Comment by YaqiWang** [comment]
@roystgnr Please let me know when this is ready for a try.

**Comment by roystgnr** [comment]
@YaqiWang - this should be in the MOOSE libMesh submodule now, if you want to try it out on a `METHOD=devel` run.  I wouldn't be hopeful that it to affect one of the stack trace sets you showed us (the PetscVector init), but the other (the FEProblemBase::outputStep) is what inspired this and I'd be very curious to know if you manage to trip any assertions.

**Comment by roystgnr** [comment]
Well, "in the MOOSE libMesh submodule" in next, anyway.  Looks like it hasn't made it to devel yet.

**Pull request merged**